### ISSUE TITLE
Softmax.apply should not use theano.tensor.nnet.softmax

### ISSUE THREAD

**Issue by rizar** [mila-udem/blocks#659]
...because the latter is numerically unstable, see discussion in #654

**Comment by mjwillson** [comment]
Posted this in #654 but noticed that was closed, and it might be relevant here too:

I noticed in theano itself there is theano.tensor.nnet.categorical_crossentropy. For the one-hot case of this uses CrossentropyCategorical1Hot, which seems to have a grad implemented, and its docs state:

    :note: In the case that the coding distribution is the output of a
           softmax, an application of this Op will probably be optimized
           away in favour of one with a C implementation.

While I don't like the sound of "probably", is there a reason not to use this on top of theano.tensor.nnet.softmax in Softmax.categorical_crossentopy ? Is there an easy way I can find out what optimisations it's made under the hood?

It'd be good to clear this stuff up anyway, either here or in theano. I'm not enough of an expert to debug it authoritatively, but for me using blocks' Softmax.categorical_crossentopy seems to cause theano to do something suspiciously slow / inefficient (turning on profiling, around 2/3 of the time is spent in Softmax and Elemwise exp, which seems way too much...)
